pkg/operator/ceph/disruption/clusterdisruption/osd.go

```diff
@@ -168,130 +182,192 @@ func (r *ReconcileClusterDisruption) reconcilePDBsForOSDs(
 			logger.Debugf("Ceph %q cluster not ready, cannot check Ceph status yet.", request.Namespace)
 			return nil
 		}
-		return errors.Wrapf(err, "could not check cluster health")
+		return errors.Wrapf(err, "failed to check cluster health")
+	}
+
+	activeDrains := len(drainingFailureDomains) > 0
+
+	// create blocking PDBs for all the OSDs that are not being drained right now
+	if activeDrains {
+		logger.Infof("pg health: %q. detected drains on %q: %v", pgHealthMsg, failureDomain, drainingFailureDomains)
+		err := r.handleActiveDrains(allFailureDomains, drainingFailureDomains, failureDomain, clusterInfo.Namespace, clean)
+		if err != nil {
+			return errors.Wrap(err, "failed to handle active drains")
+		}
 	}
+
 	if pdbStateMap.Data == nil {
 		pdbStateMap.Data = make(map[string]string)
 	}
-	_, ok := pdbStateMap.Data[disabledPDBKey]
+	_, ok := pdbStateMap.Data[drainingFailureDomain]
 	if !ok {
-		pdbStateMap.Data[disabledPDBKey] = ""
+		pdbStateMap.Data[drainingFailureDomain] = ""
 	}
-	var disabledPDBTime time.Time
-	timeString, disabledPDBTimeSet := pdbStateMap.Data[disabledPDBTimeKey]
-	if disabledPDBTimeSet {
-		disabledPDBTime, err = time.Parse(time.RFC3339, timeString)
+	var drainingFDTime time.Time
+	timeString, drainingFDTimeSet := pdbStateMap.Data[drainingFDTimeKey]
+	if drainingFDTimeSet {
+		drainingFDTime, err = time.Parse(time.RFC3339, timeString)
 		if err != nil {
-			logger.Errorf("could not parse timestamp %v. %v", disabledPDBTime, err)
-			disabledPDBTime = time.Now()
-			pdbStateMap.Data[disabledPDBTimeKey] = disabledPDBTime.Format(time.RFC3339)
+			logger.Errorf("could not parse timestamp %v. %v", drainingFDTime, err)
+			drainingFDTime = time.Now()
+			pdbStateMap.Data[drainingFDTimeKey] = drainingFDTime.Format(time.RFC3339)
 		}
 	}
-	recentlyChanged := disabledPDBTimeSet && time.Since(disabledPDBTime) < time.Minute
-	shouldChange := clean && !recentlyChanged
-	activeDrains := len(drainingFailureDomains) != 0
-	if activeDrains {
-		logger.Infof("pg health: %q. detected drains on %q: %v", pgHealthMsg, poolFailureDomain, drainingFailureDomains)
-	}
-	if shouldChange {
-		if activeDrains {
-			pdbStateMap.Data[disabledPDBKey] = drainingFailureDomains[0]
-			pdbStateMap.Data[disabledPDBTimeKey] = time.Now().Format(time.RFC3339)
-		} else {
-			pdbStateMap.Data[disabledPDBKey] = ""
-			delete(pdbStateMap.Data, disabledPDBTimeKey)
+
+	clusterHealed := !activeDrains && clean
+	if clusterHealed {
```

Comment:
For this block, what about the following structure to handle the timeout scenario?
```go
if activeDrains {
  // Reset the timestamp for tracking the timeout
  startTimeWaitingForClean = time.Now().UTC()
} else if clean {
  // reset the PDBs
} else { // !activeDrains && !clean
  // The PGs are not clean, so make sure the timestamp is set
  if startTimeWaitingForClean == nil {
    startTimeWaitingForClean = time.Now().UTC()
  } else if time.Now().UTC() - startTimeWaitingForClean > cleanTimeout {
    // The timeout has expired, now reset the PDBs
  } 
}
```

Reply:
added timeout scenario for pg health check and unit tests.